File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:

        res = []

        m, n = len(matrix), len(matrix[0])
        row_start, row_end = 0, m-1        
        col_start, col_end = 0, n-1

        while row_start <= row_end and col_start <= col_end:

            for j in range(col_start, col_end+1):
                res.append(matrix[row_start][j])
            row_start += 1
            
            for i in range(row_start, row_end+1):
                res.append(matrix[i][col_end])
            col_end -= 1

            for j in range(col_end, col_start-1, -1):
                res.append(matrix[row_end][j])
            row_end -= 1

            for i in range(row_end, row_start-1, -1):
                res.append(matrix[i][col_start])
            col_start += 1

        return res[:m*n]

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {

        int row_begin = 0;
        int row_end = matrix.length-1;
        int col_begin = 0;
        int col_end = matrix[0].length-1;

        List<Integer> result = new ArrayList<>();

        while(row_begin <= row_end && col_begin <= col_end){

            for(int j=col_begin; j<=col_end; j++){
                result.add(matrix[row_begin][j]);
            }
            row_begin++;

            for(int i=row_begin; i<=row_end; i++){
                result.add(matrix[i][col_end]);
            }
            col_end--;
            
            if (row_begin <= row_end) {
                for(int j=col_end; j>= col_begin; j--){
                    result.add(matrix[row_end][j]);
                }                        
                row_end--;
            }

            if (col_begin <= col_end) {
                for(int i=row_end; i>= row_begin; i--){
                    result.add(matrix[i][col_begin]);
                }                     
                col_begin++;              
            }
        }        

        return result;
    }
}